### ISSUE TITLE
[BUG] thermo_modify format accepts too many fields

### ISSUE THREAD

**Issue by hammondkd** [lammps/lammps#3476]
**Summary**

The `thermo_modify` command currently allows the user to specify a format string for one too many fields. Example is below.

**LAMMPS Version and Platform**

Latest develop branch (synchronized this afternoon, 10/6/2022).

**Expected Behavior**

The following input script should produce an error:
```
region simbox block 0 5 0 5 0 5
create_box 1 simbox
thermo_style custom step etotal
thermo_modify format 3 "%G"
```
The last line should try to access the third element of a two-element vector, producing an error.

**Actual Behavior**
No error message.

**Further Information, Files, and Links**

I think I know how to fix this (it appears to be a simple off-by-one error in the bounds-checking), but just in case someone says, "NO, DON'T FIX IT THAT WAY!" I am posting this here first. I'm planning to change the error-checking bounds, not resize `format_column_user`. I see in other parts of thermo_modify.cpp that there is an "invisible" field containing the volume of the box, which is why this doesn't cause a more serious problem.

I am also writing a feature (and tripped this bug while testing it) that will allow things like `thermo_modify format 3*5 "%G"` to be used to set the format of consecutive columns at the same time.

**Comment by akohlmey** [comment]
Please don't invest too much effort into playing around with format strings.

We are seriously considering to get rid of them and take advantage of `{fmt}` formatting instead.
With that we can simplify the format requests to request only "width", "precision" (if for a floating point number) and "alignment" instead of providing "raw" format strings. You can see this already in the thermo column headers that are aligned and centered(!) with the data columns for as long as you stick with the default format (and no longer condensed to the left).

With `{fmt}` you can format to a variable width with `fmt::format("{:{}}", intvalue, width);` and `fmt::format("{:{}.{}}", fpvalue, width, precision);` and you don't have to worry about doctoring the format string for 64-bit integers.

In summary, the implementation would be simpler and safer and it would be simpler to use for many users, especially non-programmers, at the expense of taking a little bit of flexibility away from the most sophisticated users.

**Comment by hammondkd** [comment]
I'm not changing the format strings at all here---it's the fact that column 3 doesn't exist that's the bug.

I have a fix in place that I'll push shortly, and it also implements the range idea. If the "%f" and such format strings go away in favor of "{fmt}" and friends, these changes should be agnostic to that.

The bug seems to be an off by TWO error: "+ 1" should have been "- 1".

**Comment by akohlmey** [comment]
please just provide the proposed bugfix here. If this is for real and addressing the right issue, it will affect other parts of the thermo code and possibly other parts of LAMMPS, too. I'd also prefer to add tests for it to the unittest programs. I have a pull request already open with fixes and improvements there (and some refactoring accessing computes/fixes/variables in various places).

**Comment by akohlmey** [comment]
when comparing `thermo_modify format` with `thermo_modify colname` it appears the fix would be:
```
  index 518deab618..df07e7c87b 100644
  --- a/src/thermo.cpp
  +++ b/src/thermo.cpp
  @@ -691,7 +691,7 @@ void Thermo::modify_params(int narg, char **arg)
               icol = -1;
             }
           }
  -        if (icol < 0 || icol >= nfield_initial + 1)
  +        if ((icol < 0) || (icol >= nfield_initial))
             error->all(FLERR, "Invalid thermo_modify format argument: {}", arg[iarg + 1]);
           format_column_user[icol] = arg[iarg + 2];
         }
```

**Comment by hammondkd** [comment]
> please just provide the proposed bugfix here. If this is for real and addressing the right issue, it will affect other parts of the thermo code and possibly other parts of LAMMPS, too. I'd also prefer to add tests for it to the unittest programs. I have a pull request already open with fixes and improvements there (and some refactoring accessing computes/fixes/variables in various places).

They should be available now in https://github.com/hammondkd/lammps/commit/3b292cd50a5c036553c790557b23a6fe8d0ffe0e; if I should submit it as a pull request, I can do that.

The bug fix is line 694 [new file line 709] of src/thermo.cpp (it seems like it should be nfield_initial, not nfield_initial - 1, but that didn't work).

**Comment by akohlmey** [comment]
Please note that `icol > nfield_initial - 1` in your commit is the same as `icol >= nfield_initial` in my change. This is the same what is also used for `thermo_modify colname` which I had tested extensively. Not sure how the second one was implemented differently, since both options were added in the same commit e3611c5d73ffe7886f4578d633a5886a7114924f
I prefer to have both comparisons implemented the same way and my variant is simpler, too.

**Comment by hammondkd** [comment]
Yes, they are the same (I implemented that before your previous post).

If you would like me to excise the "new" feature from this into another pull request, I can do that.

**Comment by akohlmey** [comment]
I just added d83d15053676b40b1e15c40402053ba75b588ec9 to PR #3475
